### PR TITLE
test(devtools-example): Add e2e test to devtools-example

### PR DESCRIPTION
**Pull request by chentong7**
## Description

This change is to config jest puppeteer in dev-example

~~~
tong@tongchen-pc:~/workspace/FluidFramework/packages/tools/devtools/devtools-example$ npm run test:jest

> @fluid-example/devtools-example@2.0.0-internal.6.2.0 test:jest
> jest --detectOpenHandles

Determining test suites to run...<i> [webpack-dev-server] Project is running at:
<i> [webpack-dev-server] Loopback: http://localhost:8231/
<i> [webpack-dev-server] On Your Network (IPv4): http://172.21.49.168:8231/
<i> [webpack-dev-server] Content not from webpack is served from '/home/tong/workspace/FluidFramework/packages/tools/devtools/devtools-example/public' directory
<i> [webpack-dev-middleware] wait until bundle finished: /
 PASS  src/test/ExampleUi.test.ts (23.048 s)
  End to end tests
    ✓ Smoke: verify test app can be launched (733 ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        23.088 s, estimated 24 s
Ran all test suites.
<i> [webpack-dev-server] Gracefully shutting down. To force exit, press ^C again. Please wait...
~~~

### CODE REVIEW

File: packages/tools/devtools/devtools-example/src/test/ui-smoke-test/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+	"extends": "@fluidframework/build-common/ts-common-config.json",
+	"compilerOptions": {
+		"module": "esnext",
+		"outDir": "../dist/ui-smoke-test",
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/src/test/ui-smoke-test/tsconfig.json:5]: The `outDir` should end up outside of `src/`, so the output files aren't seen as input files by the "regular" build the next time it runs.

That said, I even wonder if we should have a separate build for the tests in this package... the unit tests in `src/test` already get built as part of the single build we do right now for all the source code, and for an example app I think it's probably fine to have the jest test be part of that too. @Josmithr, thoughts? It means the single tsconfig gets chunky. needing all the types specified here, but I think it's ok for a sample app.

And if we do, the whole package should probably be ESM (since it's not library), and the single build be a `module: "esnext"` one.

File: packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		await page.goto(globals.PATH, { waitUntil: "load" });
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-return
+		await page.waitForFunction(() => globalThis.fluidStarted);
+		await page.waitForSelector("textarea");
+	});
+
+	it("Smoke: verify test app can be launched", async () => {
+		// Verify by checking for text area associated with the SharedString.
+		const textArea = await getTextFormValue("Enter text here.");
+		expect(textArea).toEqual("Enter text here.");
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts:54]: The way the test works feels a bit wonky... shouldn't we just be doing `page.waitForSelector()` on an id we give to the element we care about, and then check that the text in it is what we expect? Here the helper `getTextFromValue` seems like it would only ever return the expected value.

File: packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+				}),
+			/* check: */ (actualValue) => actualValue === expectedValue,
+			/* defaultValue: */ "",
+		);
+	}
+
+	beforeAll(async () => {
+		// Wait for the page to load first before running any tests
+		// so this time isn't attributed to the first test
+		await page.goto(globals.PATH, { waitUntil: "load", timeout: 0 });
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts:41]: We have a `page.goto()` in `beforeEach()` as well, so _some_ navigation time still gets attributed to each test. Is this to handle the first load, which might be longer because of webpack dev server / similar?

File: packages/tools/devtools/devtools-example/src/index.tsx [hunk 4]
```diff
@@ -29,6 +29,9 @@ ReactDOM.render(<DevtoolsView />, devtoolsElement, () => {
 	console.log("Devtools UI rendered!");
 });
 
+// Setting "fluidStarted" is just for our test automation
+globalThis.fluidStarted = true;
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/src/index.tsx:33]: Should this maybe go in the callback to the `ReactDOM.render()` call that renders the app? So it's set once we really have everything in the DOM. In here, this could technically get set to `true` before the app is visible.

File: packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts:1]: This is just and e2e test for the sample app, not really a test for the extension, right? Let's rename the file to reflect that.

File: packages/tools/devtools/devtools-example/jest-puppeteer.config.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+		usedPortAction: "error",
+	},
+	launch: {
+		args: [
+			// https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md#setting-up-chrome-linux-sandbox
+			"--no-sandbox",
+			"--disable-setuid-sandbox",
+
+			// Ensure our extension is loaded into the browser environment
+			"--load-extension=./dist/bundle",
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/jest-puppeteer.config.js:20]: This package doesn't have access to the extension, this shouldn't be necessary here.

File: packages/tools/devtools/devtools-example/jest.config.js [hunk 6]
```diff
@@ -2,16 +2,24 @@
...
-	roots: ["<rootDir>/src"],
+	preset: "jest-puppeteer",
+	globals: {
+		PATH: `http://localhost:${mappedPort}`,
+	},
+	testMatch: ["**/ui-smoke-test/?(*.)+(spec|test).[t]s"],
 	transform: {
-		"^.+\\.tsx?$": "ts-jest",
+		"^.+\\.test.ts?$": "ts-jest",
 	},
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/jest.config.js:20]: I think this is making the existing unit test in `src/test/` to stop running...?

File: packages/tools/devtools/devtools-example/jest.config.js [hunk 7]
```diff
@@ -2,16 +2,24 @@
...
-	testRegex: "(/__tests__/.*|(\\.|/)(test|spec))\\.(ts|tsx)?$",
 	testPathIgnorePatterns: ["/node_modules/", "dist"],
-	moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json", "node"],
-	coveragePathIgnorePatterns: ["/node_modules/", "/src/test/"],
-	testEnvironment: "jsdom",
+	// While we still have transitive dependencies on 'uuid<9.0.0', force the CJS entry point:
+	// See: https://stackoverflow.com/questions/73203367/jest-syntaxerror-unexpected-token-export-with-uuid-library
+	moduleNameMapper: { "^uuid$": "uuid" },
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-example/jest.config.js:24]: Can't quite tell why the changes here; let's just make sure they're not affecting the existing unit test.

**Comment by alexvy86** [comment]
I think I'd also rename the PR to "Add e2e test to devtools-example" or something a bit more descriptive of the final goal of the PR.

**Comment by chentong7** [comment]
> I think I'd also rename the PR to "Add e2e test to devtools-example" or something a bit more descriptive of the final goal of the PR.

Renamed!

File: packages/tools/devtools/devtools-example/jest.config.js (hunk 7 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/jest.config.js:28]: Yes. It's required.

File: packages/tools/devtools/devtools-example/jest.config.js (hunk 6 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/jest.config.js]: Fixed

File: packages/tools/devtools/devtools-example/jest-puppeteer.config.js (hunk 5 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/jest-puppeteer.config.js]: Removed

File: packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts:1]: Renamed

File: packages/tools/devtools/devtools-example/src/index.tsx (hunk 4 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/index.tsx]: Resolved

File: packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts (hunk 3 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts]: This is for open a new page before each test

File: packages/tools/devtools/devtools-example/.eslintrc.js [hunk 8]
```diff
@@ -31,6 +32,10 @@ module.exports = {
 			rules: {
 				"import/no-nodejs-modules": "off",
 				"unicorn/prefer-module": "off",
+				"unicorn/prefer-query-selector": "off",
+				"@typescript-eslint/no-unsafe-argument": "off",
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/.eslintrc.js:36]: Do we need to disable these typescript-eslint rules? These would be good for us to abide by in our tests too.

File: packages/tools/devtools/devtools-example/jest.config.js [hunk 9]
```diff
@@ -2,16 +2,28 @@
...
 module.exports = {
-	preset: "ts-jest",
+	preset: "jest-puppeteer",
 	roots: ["<rootDir>/src"],
+	globals: {
+		PATH: `http://localhost:${mappedPort}`,
+	},
 	transform: {
-		"^.+\\.tsx?$": "ts-jest",
+		"^.+\\.(ts|tsx)?$": "ts-jest",
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/jest.config.js:19]: This change isn't correct. It will match files with no extension, in addition to ts and tsx files. What was there before was correct and would match on both ts and tsx files.

File: packages/tools/devtools/devtools-example/jest.config.js [hunk 10]
```diff
@@ -2,16 +2,28 @@
...
 	},
-	testRegex: "(/__tests__/.*|(\\.|/)(test|spec))\\.(ts|tsx)?$",
+	testTimeout: 30_000,
+	testRegex: "(/__tests__/.*|(\\.|/)(test|spec))\\.ts?$",
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/jest.config.js:22]:
```suggestion
	testRegex: "(/__tests__/.*|(\\.|/)(test|spec))\\.tsx?$",
```

File: packages/tools/devtools/devtools-example/jest.config.js (hunk 9 above)
Comment (Josmithr) [reply packages/tools/devtools/devtools-example/jest.config.js:19]:
```suggestion
		"^.+\\.tsx?$": "ts-jest",
```

File: packages/tools/devtools/devtools-example/jest.config.js [hunk 11]
```diff
@@ -2,16 +2,28 @@
...
 	testPathIgnorePatterns: ["/node_modules/", "dist"],
 	moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json", "node"],
 	coveragePathIgnorePatterns: ["/node_modules/", "/src/test/"],
-	testEnvironment: "jsdom",
+	// While we still have transitive dependencies on 'uuid<9.0.0', force the CJS entry point:
+	// See: https://stackoverflow.com/questions/73203367/jest-syntaxerror-unexpected-token-export-with-uuid-library
+	moduleNameMapper: { "^uuid$": "uuid" },
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/jest.config.js:28]: Is this needed? This package should be using common js already.

File: packages/tools/devtools/devtools-example/tsconfig.json [hunk 12]
```diff
@@ -3,9 +3,10 @@
 	"compilerOptions": {
 		"rootDir": "src",
 		"outDir": "dist",
+		"module": "ESNext",
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/tsconfig.json:6]: Not opposed to this change, just curious what it is needed for?

File: packages/tools/devtools/devtools-example/jest.config.js (hunk 11 above)
Comment (Josmithr) [reply packages/tools/devtools/devtools-example/jest.config.js:28]: Sorry, I see the change to ESNext here as well. I left a question about that, but I'll close this one.

File: packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts (hunk 2 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/test/ui-smoke-test/ChromeBrowserDevtool.test.ts]: Right! Adding wait for to ensure app view is fully rendered.

File: packages/tools/devtools/devtools-example/src/test/ui-smoke-test/tsconfig.json (hunk 1 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/src/test/ui-smoke-test/tsconfig.json]: I see! I synced-up with Josh offline. We are planning to disable the `CounterWidget` test and only runs e2e tests because its test environment is conflict with jest pupeteer, and just keep `src/test` to have built jest tests.

File: packages/tools/devtools/devtools-example/tsconfig.json (hunk 12 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/tsconfig.json:6]: I probably can refer to @alexvy86's comments before. The test-until is not a lib as test lib. We should need to define it as ESM

File: packages/tools/devtools/devtools-example/.eslintrc.js (hunk 8 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/.eslintrc.js]: Removed

File: packages/tools/devtools/devtools-example/tsconfig.json (hunk 12 above)
Comment (Josmithr) [reply packages/tools/devtools/devtools-example/tsconfig.json:6]: Ah, okay. Makes sense, and sounds good!

File: packages/tools/devtools/devtools-example/.eslintrc.js [hunk 13]
```diff
@@ -31,6 +32,8 @@ module.exports = {
 			rules: {
 				"import/no-nodejs-modules": "off",
 				"unicorn/prefer-module": "off",
+				"unicorn/prefer-query-selector": "off",
+				"@typescript-eslint/no-unsafe-return": "off",
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/.eslintrc.js:36]: Can we restore this rule also?

File: packages/tools/devtools/devtools-example/jest.config.js [hunk 14]
```diff
@@ -2,16 +2,28 @@
...
 module.exports = {
-	preset: "ts-jest",
+	preset: "jest-puppeteer",
 	roots: ["<rootDir>/src"],
+	globals: {
+		PATH: `http://localhost:${mappedPort}`,
+	},
 	transform: {
-		"^.+\\.tsx?$": "ts-jest",
+		"^.+\\.ts?$": "ts-jest",
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/jest.config.js:19]:
```suggestion
		"^.+\\.tsx?$": "ts-jest",
```

Comment (chentong7) [reply packages/tools/devtools/devtools-example/jest.config.js:19]: Our e2e test is .ts

File: packages/tools/devtools/devtools-example/package.json [hunk 15]
```diff
@@ -89,12 +98,15 @@
...
 		"jest": "^29.6.2",
+		"jest-dev-server": "^4.3.0",
 		"jest-environment-jsdom": "^29.6.2",
 		"jest-junit": "^10.0.0",
 		"pm2": "^5.3.0",
 		"prettier": "~2.6.2",
+		"puppeteer": "^17.1.3",
 		"rimraf": "^4.4.0",
-		"start-server-and-test": "^1.11.7",
+		"sinon": "^7.4.2",
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/package.json:108]: Are we using sinon / sinon-chrome for anything? If not, let's remove these and the corresponding @types imports

File: packages/tools/devtools/devtools-example/package.json [hunk 16]
```diff
@@ -72,14 +72,23 @@
...
 		"@testing-library/dom": "^8.2.0",
 		"@testing-library/jest-dom": "^5.16.5",
 		"@testing-library/react": "^12.0.0",
 		"@testing-library/user-event": "^14.4.3",
+		"@types/chrome": "0.0.232",
+		"@types/expect-puppeteer": "2.2.1",
 		"@types/jest": "29.5.3",
+		"@types/jest-environment-puppeteer": "2.2.0",
+		"@types/jsdom": "^21.1.1",
+		"@types/jsdom-global": "^3.0.4",
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-example/package.json:86]: Do we need the jsdom dependencies? We're using the puppeteer environment now, so I don't think they're needed?

**Review by Josmithr — APPROVED** [review]
The code changes look good, but I think we need to do an audit of the new dependencies added to the package - I believe many of them are unneeded.

Also, I remember there were issues with the CounterWidget component tests. Are those no longer broken? Or do they need to be deleted still?

File: packages/tools/devtools/devtools-example/.eslintrc.js (hunk 13 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/.eslintrc.js:36]: Yes, removed

File: packages/tools/devtools/devtools-example/package.json (hunk 16 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/package.json:86]: Yes, we can remove the jsdom dependencies as well. Then should I also delete CounterWidget.test.tsx file as well?

File: packages/tools/devtools/devtools-example/package.json (hunk 15 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/package.json]: Cleaned up

File: packages/tools/devtools/devtools-example/package.json (hunk 16 above)
Comment (chentong7) [reply packages/tools/devtools/devtools-example/package.json]: Cleaned up

File: packages/tools/devtools/devtools-example/package.json [hunk 17]
```diff
@@ -72,14 +72,23 @@
 		"@fluidframework/build-common": "^2.0.0",
 		"@fluidframework/build-tools": "^0.22.0",
 		"@fluidframework/eslint-config-fluid": "^2.0.0",
+		"@fluidframework/test-utils": "workspace:~",
 		"@microsoft/api-extractor": "^7.34.4",
 		"@testing-library/dom": "^8.2.0",
 		"@testing-library/jest-dom": "^5.16.5",
 		"@testing-library/react": "^12.0.0",
 		"@testing-library/user-event": "^14.4.3",
+		"@types/chrome": "0.0.232",
```
[the remark this answers is not in the document]
Comment (chentong7) [reply packages/tools/devtools/devtools-example/package.json]: Cleaned up
